fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java

```diff
@@ -0,0 +1,458 @@
+    }
+
+    // PostVisitor is used to rewrite the current node, the visit function must satisfy:
+    // 1. return RewriteInfo.NOT_REWRITE to short-circuit the later bottom-up rewrite operation.
+    // 2. put new-created OptExpression into RewriteInfo by invoke RewriteInfo.setOp if the new
+    // OptExpression is created.
+    // 3. return input RewriteInfo as return value if you want to rewrite upper nodes.
+    private class PostVisitor extends OptExpressionVisitor<RewriteInfo, RewriteInfo> {
+
+        // Default visit method do nothing but just pass the RewriteInfo to its parent
+        @Override
+        public RewriteInfo visit(OptExpression optExpression, RewriteInfo rewriteInfo) {
+            if (rewriteInfo != RewriteInfo.NOT_REWRITE) {
+                rewriteInfo.setOp(optExpression);
+            }
+            return rewriteInfo;
+        }
+
+        private CallOperator genAggregation(CallOperator origin) {
+            ScalarOperator arg = origin.getChild(0);
+            Function fn = Expr.getBuiltinFunction(origin.getFunction().getFunctionName().getFunction(),
+                    new Type[] {arg.getType()}, Function.CompareMode.IS_NONSTRICT_SUPERTYPE_OF);
+
+            Preconditions.checkState(fn instanceof AggregateFunction);
+            if (arg.getType().isDecimalOfAnyVersion()) {
+                fn = DecimalV3FunctionAnalyzer.rectifyAggregationFunction((AggregateFunction) fn, arg.getType(),
+                        origin.getType());
+            }
+
+            return new CallOperator(fn.getFunctionName().getFunction(), fn.getReturnType(),
+                    Lists.newArrayList(arg.clone()), fn);
+        }
+
+        @Override
+        public RewriteInfo visitLogicalTableScan(OptExpression optExpression, RewriteInfo rewriteInfo) {
+            AggregatePushDownContext ctx = rewriteInfo.getCtx();
+            if (ctx.groupBys.isEmpty() || ctx.aggregations.isEmpty()) {
+                return RewriteInfo.NOT_REWRITE;
+            }
+            List<ColumnRefOperator> groupBys = ctx.groupBys.values().stream().map(ScalarOperator::getUsedColumns)
+                    .flatMap(colSet -> colSet.getStream().mapToObj(factory::getColumnRef)).distinct()
+                    .collect(Collectors.toList());
+
+            LogicalScanOperator scanOp = optExpression.getOp().cast();
+            ColumnRefSet scanOutputColRefSet = new ColumnRefSet(scanOp.getOutputColumns());
+            Map<CallOperator, ColumnRefOperator> uniqueAggregations = Maps.newHashMap();
+            Map<ColumnRefOperator, ColumnRefOperator> remapping = Maps.newHashMap();
+            Preconditions.checkArgument(scanOutputColRefSet.containsAll(new ColumnRefSet(groupBys)));
+
+            for (Map.Entry<ColumnRefOperator, CallOperator> entry : ctx.aggregations.entrySet()) {
+                CallOperator aggCall = entry.getValue();
+                Preconditions.checkArgument(aggCall.getChildren().size() == 1 && aggCall.getChild(0).isColumnRef());
```

Comment:
I think you can't promise the check always true

Reply:
at present,  only sum(column) over (...) is processed, so the promise can be ensured.